src/eval/mod.rs

```diff
@@ -1184,6 +1184,51 @@ impl Eval for ast::Closure {
     }
 }
 
+impl ast::Pattern {
+    // Destruct the given value into the pattern.
+    pub fn define(&self, vm: &mut Vm, value: Value) -> SourceResult<Value> {
+        match self.bindings() {
+            ast::PatternKind::Ident(ident) => {
+                vm.define(ident, value);
+                Ok(Value::None)
+            }
+            ast::PatternKind::Destructure(pattern) => {
+                let Ok(value) = value.clone().cast::<Array>() else {
+                    bail!(self.span(), "cannot destructure {} into an array", value.type_name());
+                };
+
+                let mut i = 0;
+                for p in &pattern {
+                    match p {
+                        ast::DestructuringKind::Ident(ident) => {
+                            let Ok(v) = value.at(i) else {
+                                bail!(ident.span(), "not enough elements to destructure");
+                            };
+                            vm.define(ident.clone(), v.clone());
+                            i += 1;
+                        }
+                        ast::DestructuringKind::Sink(ident) => {
+                            let sink_size = i64::try_from(value.len()).unwrap()
+                                - i64::try_from(pattern.len()).unwrap()
+                                + 1;
+                            let Ok(v) = value.slice(i, Some(i + sink_size)) else {
```

Comment:
note that `slice` allows negative indices to index from the back. does this interfere with negative sink size?